### PR TITLE
Adding URI SAN support

### PR DESCRIPTION
**Pull request by mweissbacher**

### CODE REVIEW

**Comment by mweissbacher** [comment]
This PR depends on Go 1.11

File: pkix/csr.go [hunk 1]
```diff
@@ -69,8 +70,26 @@ func ParseAndValidateIPs(ipList string) (res []net.IP, err error) {
...
+	if len(uriList) > 0 {
+		uris := strings.Split(uriList, ",")
+		for _, uri := range uris {
+			parsedURI, err := url.Parse(uri)
+			if err != nil {
+				parsedURI = nil
+			}
+			if parsedURI == nil {
+				return nil, fmt.Errorf("Invalid URI: %s", uri)
+			}
```
Comment (mcpherrinm) [inline pkix/csr.go]: url.Parse allows relative URLs, which don't make sense in this context.  We could do something like this:

```suggestion
			}
			if !parsedURI.IsAbs() {
				return nil, fmt.Errorf("Invalid URI: %s", uri)
			}
```

Comment (mcpherrinm) [reply pkix/csr.go]: And a test to verify '/' and such are rejected

**Comment by mcpherrinm** [comment]
I tested a bit, and it seems like the CSRs work, but we'll need to add support to `certstrap sign` too.

also, it's pretty weird to put a URI in the Common Name.  I think that likely isn't what anyone will intend.  Maybe we should make -common-name required if no --domain flag is passed too

**Comment by mweissbacher** [comment]
> also, it's pretty weird to put a URI in the Common Name. I think that likely isn't what anyone will intend. Maybe we should make -common-name required if no --domain flag is passed too

How about IPs? We could just remove URIs (and IPs?) from the naming switch statement in `cmd/request_cert.go` and it'll fail automatically if no domain or common name are provided.

File: README.md [hunk 2]
```diff
@@ -66,9 +66,9 @@ Created out/Alice.key
 Created out/Alice.csr
 ```
 
-certstrap requires at least one of `-common-name`, `-ip`, or `-domain` flags to be set in order to generate a certificate signing request.  The CN for the certificate will be found from these fields.
+certstrap requires at least one of `-common-name`, `-ip`, `-domain`, or `-uri` flags to be set in order to generate a certificate signing request.  The CN for the certificate will be found from these fields.
```
Comment (mcpherrinm) [inline README.md]: This line isn't quite accurate.  We require at least one of -common-name or -domain now

**Comment by mcpherrinm** [comment]
One minor nit in the README.md, otherwise looks good now

**Pull request merged**